ingest/src/main/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClient.java

```diff
@@ -112,15 +160,50 @@ public IngestionResult ingestFromStream(StreamSourceInfo streamSourceInfo, Inges
         streamSourceInfo.validate();
         ingestionProperties.validate();
 
-        if (streamSourceInfo.isLeaveOpen()) {
-            throw new UnsupportedOperationException("LeaveOpen can't be true in ManagedStreamingIngestClient");
+        UUID sourceId = streamSourceInfo.getSourceId();
+        if (sourceId == null) {
+            sourceId = UUID.randomUUID();
+        }
+
+        byte[] streamingBytes;
+
+        try {
+            streamingBytes = IOUtils.readNBytes(streamSourceInfo.getStream(), MAX_STREAMING_SIZE_BYTES + 1);
+        } catch (IOException e) {
+            throw new IngestionClientException("Failed to read from stream.", e);
         }
-        streamSourceInfo.setLeaveOpen(true);
 
+        ByteArrayInputStream byteArrayStream = new ByteArrayInputStream(streamingBytes);
+
+        if (streamingBytes.length > MAX_STREAMING_SIZE_BYTES) {
+            log.info("Stream size is greater than max streaming size (%d bytes). Falling back to queued.");
+            StreamSourceInfo managedSourceInfo = new StreamSourceInfo(new SequenceInputStream(byteArrayStream, streamSourceInfo.getStream()),
+                    streamSourceInfo.isLeaveOpen(), sourceId);
+            managedSourceInfo.setCompressionType(streamSourceInfo.getCompressionType());
+            return queuedIngestClient.ingestFromStream(managedSourceInfo, ingestionProperties);
+        }
+
+        if (!streamSourceInfo.isLeaveOpen()) {
+            // From this point we don't need the original stream anymore, we cached it
+            try {
+                streamSourceInfo.getStream().close();
+            } catch (IOException e) {
+                log.warn("Failed to close stream", e);
+            }
+        }
+
+        StreamSourceInfo managedSourceInfo = new StreamSourceInfo(byteArrayStream, true, sourceId);
+        managedSourceInfo.setCompressionType(streamSourceInfo.getCompressionType());
+
+        ExponentialRetry retry = new ExponentialRetry(exponentialRetryTemplate.getMaxAttempts(), exponentialRetryTemplate.getSleepBaseSecs(),
+                exponentialRetryTemplate.getMaxJitterSecs());
         try {
-            for (int i = 0; i < MAX_RETRY_CALLS; i++) {
+            while (retry.shouldTry()) {
                 try {
-                    return streamingIngestClient.ingestFromStream(streamSourceInfo, ingestionProperties);
+                    log.info("Streaming ingest attempt {}", retry.getCurrentAttempt());
+                    String clientRequestId = String.format("KJC.executeManagedStreamingIngest;%s;%d", sourceId,
```

Comment:
I'm sure you already are, but just calling out that we should be deliberate in our `clientRequestId` including when we change approaches mid-ingest (at first glance I'd expect that the CRID shouldn't change).

Reply:
If you mean when we move to queued - we don't use ClientRequestId in queued, not sure how that would work

Reply:
Right, great point :-)

But doesn't this approach call methods in both ManagedStreaming and Streaming classes? I'm sure it is, but just making sure our CRID is consistent.

Reply:
As far as I tested it does work, if you think there's a flow I missed I'd be happy to know